### ISSUE TITLE
More flexible Yaml with Matrix

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#902]
I would like to solve issues #865 #874 and #696 by proposing a more flexible yaml format. The initial yaml was inspired by Travis with some minor modifications to better align with Docker.

Below is my proposal. It is an evolution of the existing yaml with inspiration from `docker compose`. It also takes the new (proposed) plugin system into account.

Example:

```yaml
build:
  image: python
  envs:
    - foo=bar
    - baz=boo
  cmds:
    - go build
    - go test

compose:
  redis:
    image: redis

  postgres:
    image: postgres

publish:
  azure:
    image: azure
    args:
      token: 123123121
      region: east-1
```

This should work well for #696 #874 since it would allow multiple named notifications or deployments for the same plugin:

```yaml
notify:
  proj_managers:
    image: slack
    when:
      branch: prod
    args:
      channel: project_managers
      username: drone

  devs:
    image: slack
    args:
      channel: devs
      username: drone
```

The above example may be contrived, however, it would allow us to notify two different groups, to two different slack channels, for two different events. Notice the `slack` is the Docker image (plugin) and `devs` and `proj_managers` are just labels that we use to differentiate / label the notification.

I believe this could also satisfy #865 since service declarations could be more verbose, similar to `docker compose`:

```yaml
compose:
  redis:
    image: redis
    volume: {{awesome_volume}}:/var/lib/pgsql/data
```

The downside here is that things are a bit more verbose than before. So I wanted to get everyone's thoughts and see if it makes sense to sacrifice brevity in favor of flexibility.

cc @donny-dont / @dave-tucker / @dvigueras / @vaijab / @epipho / @gdborton

**Comment by vaijab** [comment]
Looks much more powerful and flexible. However I see one potential issue - yaml hashes are not sorted, which could cause undesired behaviour.

**Comment by bradrydzewski** [comment]
maybe we could add an optional "weight" or "order" attribute and sort the values prior to execution?

**Comment by epipho** [comment]
I like the order attribute. Long term we could allow sections with identical order values to execute concurrently.

**Comment by gdborton** [comment]
I think this is great, the added verbosity shouldn't be an issue as long as it is well documented.

**Comment by dave-tucker** [comment]
:+1: 
Providing we have access to specify the parameters to be used by "docker run" for both services and the main test container, I think this will meet the use case described in #865.

As `docker-compose` was mentioned, I think it might be a nice feature to be able to defer to compose for services. E.g, start these services described in `docker-compose.yml`
This might help manage verbosity and avoid overlap.

Perhaps:
```yaml
compose:
   from: "docker-compose.yml"
```

**Comment by bradrydzewski** [comment]
@dave-tucker yes, I think we can definitely import the `docker-compose.yml` file. I'm guessing that we would want to launch supporting containers only. For example, if I was testing Drone I wouldn't want to build and run a Drone container. So could we ignore anything in `docker-compose.yml` where a `build` attribute is provided?

I'll need to investigate further, but I wonder if we could allow you to (optionally) embed the `.drone.yml` in the `docker-compose.yml` file separated by `---`. Something like this would be valid yaml: 

```
web:
  build: .
  command: python app.py
  ports:
   - "5000:5000"
  volumes:
   - .:/code
  links:
   - redis
redis:
  image: redis

---

build:
  image: go
  commands:
    - go build
    - go test

notify:
  slack:
    image: slack
```

**Comment by vaijab** [comment]
> Maybe we could add an optional "weight" or "order" attribute and sort the values prior to execution?

+1 for adding `order` attribute and the idea of being able to run `deploy` or `publish` tasks in parallel.

**Comment by donny-dont** [comment]
Not sure if this is the best idea. But what if you had the ordering in the YAML?

```yaml
deploy: 
  - 
    assets: 
      args: 
        bucket: my-assets
        from: /path/on/local
        secret: 3213212
        to: /path/on/server
        token: 1231231
      image: s3
    website: 
      args: 
        from: /path/on/local
        to: /path/on/server
      image: rsync
```

That way its implicit rather than explicit.

**Comment by bradrydzewski** [comment]
I have a functioning prototype (command line only) of the above format with matrix builds. Before we settle on a format, I wanted to float one last idea.

We could use chronological ordering in the yaml file. The ordering would be preserved by using the yaml parser's `MapSlice` type. We could also eliminate the `compose`, `notify`, `publish` and `deploy` sections completely.

It would look something like this:

```yaml
git:
  depth: 1
  path: github.com/drone/drone

build:
  image: go
  script:
    - go build
    - go test

heroku_dev:
  image: heroku
  app: foo
  when:
    branch: master

heroku_prod:
  image: heroku
  app: foo_prod
  when:
    branch: prod

slack: # we could optionally infer the `image` from the section name
  channel: dev
  user: drone
  when:
    branch: master
```

So how do we identify a notification vs a deploy vs publish? We could include meta-data in the image to signify the type of plugin. This metadata could, for example, be embedded in the image's Dockerfile as an environment variable and retrieved using the `docker inspect` command (ie `ENV PLUGIN_TYPE=notify`). 

I could also see this being useful for versioning and backward compatibility. Every plugin receives JSON-formatted payload (example: https://github.com/drone-plugins/drone-slack#docker). We may want to change this payload over time, so getting some sort of version meta-data could help maintain backward compatibility. For example, a plugin image could include `ENV PLUGIN_VERSION=1.0`.

There is some precedence for using environment variables in creative ways with Docker. The swarm library uses environment variables to co-locate two containers on the same machine, for example:

> docker run -d -P --name redmon -e affinity:container!=redis vieux/redmon

The pros of this approach is that we've reduced a yaml to a set of steps that follow the exact same format. We could essentially just loop through our steps and execute. That format would be defined as follows in Go code:

```Go
type Step struct {
	Name        string   `json:"name"`
	Image       string   `json:"image"`
	Environment []string `json:"environment"`
	Volumes     []string `json:"volumes"`
	Hostname    string   `json:"hostname"`
	Privileged  bool     `json:"privileged"`
	Daemon      bool     `json:"privileged"`
	Net         string   `json:"net"`

	// Config represents the unique configuration details
	// for each plugin.
	Config map[string]interface{} `yaml:"config,inline" json:"config"`
}
```

The con is that every plugin must generate meta-data via the Dockerfile. This would not apply to build images (ie go, ruby, etc) but would apply to any publish, deploy or notify image.

**Comment by bradrydzewski** [comment]
so I guess it boils down to:

* do we want a flat list of steps, as defined here: https://github.com/drone/drone/issues/902#issuecomment-77115839
* or more of a hierarchy, as defined here: https://github.com/drone/drone/issues/902#issue-59364897

**Comment by JeanMertz** [comment]
> We could also eliminate the compose, notify, publish and deploy sections completely.

<del>This example looks less clear to me than the original proposed one. I liked the fact that different *jobs to be done* of the build are split into different yaml sections.</del>

The above was my original comment before reading on and seeing the potential of this new proposal. Everything is dockerized, you are no longer confined to specific types of plugins, just Docker containers that start with a given payload, and then do their thing.

I like it.

**Comment by danharvey** [comment]
You would probably still need the weights to know the dependency order if you were to run them in parallel? or maybe that's for later.

Also this is now basically docker-compose, but starting these from a known git clone and with a bit more config semantics. So can that be reused somehow?

**Comment by dave-tucker** [comment]
Personally, I like the hierarchy but I'd like to find a way to defer the compose stage to `docker-compose`
Compose would give some added flexibility in the timing and ordering in which the service containers are started which is likely to be something that would be needed in prod to...

So, expanding, on my earlier idea...

```yaml
compose:
  from-file: docker-compose.yml
  services:
    - redis
    - memcached
```

This would allow you to start `memcached` and `redis` from `docker-compose`, meaning that your application could be excluded if it were in compose too...

Alternatively you could use the format described in https://github.com/drone/drone/issues/902#issue-59364897 if you didn't want to use compose.

The flat steps are ok, but ordering build steps top down makes the user have to think very carefully about the ordering of the steps...
I like some of the concepts of Puppet (e.g `requires` and `notify`) and letting Drone work out what order to do things in vs. me having to explicitly define it.

**Comment by bradrydzewski** [comment]
@dave-tucker I'm not married to either approach so it is good to get feedback.

The services section will need additional configuration options beyond the string array of image names, unfortunately. The official mysql image immediately comes to mind, which cannot start with username and password environment variables:

```
compose:
  mysql:
    image: library/mysql:5.5
    environment:
      - mysql_password=pa$$word
```

I absolutely support importing the `docker-compose.yml` file as you've described. I do think there are some open design questions that we'd need to work through. For example, the ability to assign container names in the `docker-compose.yml` file could conflict with Drone's ability to run multiple builds at the same time (due to container names already being used).

Let's make sure we choose a yaml format that enables `docker-compose` integration, but let's plan on using a separate issue (#906) to hammer out the exact implementation details.

**Comment by bradrydzewski** [comment]
I've spent a day thinking about it and evaluated the feedback.

Lets keep the category labels (ie notify, deploy, etc) for the time being as proposed in my initial comment [here](https://github.com/drone/drone/issues/902#issue-59364897). It makes parsing and ordering much more simple. I believe can always remove category labels in the future, while maintaining some level of backward compatibility.

Since each section of the yaml uses an identical structure for defining a step we can greatly simplify how this is represented and executed in Go:

```Go
type Config struct {
	Clone Step
	Build Step

	Compose map[string]Step
	Publish map[string]Step
	Deploy  map[string]Step
	Notify  map[string]Step

	Matrix map[string][]string
}

type Step struct {
	Image       string
	Environment []string
	Volumes     []string
	Hostname    string
	Privileged  bool
	Net         string

	// Config represents the unique configuration details
	// for each build step.
	Config map[string]interface{} `yaml:"config,inline"`
}
```

I have a (barely) functioning prototype in the drone/drone-cli repository in the matrix branch. It uses the above yaml formats, works with matrix builds, and uses a new dockerclient (from the Docker team) that is better tested and provides better error messaging.

I'm going to spend the rest of the week trying to productionalize this code so that it can be merged into mainline as soon as possible.

**Comment by hd-deman** [comment]
Just a note: docker have [extend](https://github.com/docker/compose/blob/master/docs/yml.md#extends) feature. so there may be cases when drone compose file will be extended from base compose file.

**Comment by hd-deman** [comment]
Is new format make possible to connect to test container from the service?
case: I have selenium service container, so when test connect to selenium service, selenium must have access to test container where project started.

**Comment by thomasf** [comment]
I'm sorry because I'm having trouble keeping up with the changes everywhere.. I'm not sure about the relations between the various kinds of drone plugins and the specs being discussed here, I get the feeling that these are at least closely related.

The use case I'm thinking of is specialized containers for individual build steps. The example is that after my postgresql container is up and running I want to execute the db migrations from my liquibase container (https://github.com/alkasir/docker-liquibase) and then proceed to run the build in another one.

**Comment by arthurgeek** [comment]
> yes, I think we can definitely import the docker-compose.yml file. I'm guessing that we would want to launch supporting containers only? For example, when testing Drone I want to start a Postgres and MySql container, however, I don't want to build and start the Drone container. So could we ignore anything in docker-compose.yml where a build attribute is provided?

This new format is great, but I don't see the point in ignoring anything with `build` attribute.

I have a private registry, but I can't use this private registry image name in `.drone.yml` because it would lead to a race condition: a push to github leads to two hooks: one to the private registry, which would build the new docker image, and another one to drone, which would still pull the old image from my private registry.

If drone is able to build the docker image instead of just pulling from a pre-built one, this would solve this problem. I would like it to fully use the `docker-compose` file. Also, with `compose` extend option, we could point `.drone.yml` to a `docker-compose-test.yml`, which would only list things needed for testing purposes.

**Comment by bradrydzewski** [comment]
The development is not yet complete, but given the relatively poor (almost non-existent) documentation for the 0.3 yaml I thought I'd try something new -- write the documentation first and use it as a spec. I will start coding to the spec discussed in this thread and defined in these docs:
http://readme.drone.io/docs/build/

Feedback welcome

**Comment by donny-dont** [comment]
@bradrydzewski what about the TOML stuff or environment variables for setting global options. Are those still a thing?

**Comment by bradrydzewski** [comment]
@donny-dont still undecided on final implementation, but thinking of forwarding all environment variables that are prefixed with `PLUGIN_` to the build.

For example, if you had a global npm username and password, you might set the following environment variables on the host machine, which in turn get passed to the build:

```
PLUGIN_NPM_USER=
PLUGIN_NPM_PASS=
```

These will be defined on a per-plugin bases. Meaning, it will be up to the plugin to define, document and query for these environment variables.

This means you can add your own plugin, with your own global variables, and not have to update the Drone source code to know about those globals.

**Comment by bradrydzewski** [comment]
since this is implemented, and 0.4-alpha.1 snapshot will be out next week, I'm going to close this
